File: ruby/bfs.rb
class Person
  attr_accessor :name, :friends, :visited

  def initialize(name)
    @name = name
    @friends = []
    @visited = false
  end

  def add_friend(friend)
    @friends << friend
  end

  def display_network
    # keep track of every node we ever visit, so we can reset their
    # ‘visited’ attribute back to false after our algorithm is complete
    to_reset = [self]

    # Create the queue. It starts out containing the root vertex
    queue = [self]
    self.visited = true

    while queue.any?
      current_vertetx = queue.shift
      puts current_vertetx.name

      # We add all adjacent vertices of the current vertex to the queue
      current_vertetx.friends.each do | friend |
        if !friend.visited
          to_reset << friend
          queue << friend
          friend.visited = true
        end
      end
    end

    # After the algorithm is complete, we reset each node’s visited attribute
    to_reset.each do | node |
      node.visited = false
    end
  end
end

File: javaScript/problems/algorithms/bfs.ts
type Graph = { [key: string]: string[] };

const graph: Graph = {
  A: ['B', 'C'],
  B: ['A', 'D', 'E'],
  C: ['A', 'F'],
  D: ['B'],
  E: ['B', 'F'],
  F: ['C', 'E']
};

/**
 *
 * BFS Example, implement BFS and log every node at each level
 */
const bfs = (graph: Graph, start: string): string[] => {
  if (!graph[start]) return [];

  let res: string[] = [];
  const queue: string[] = [start];
  const visited: Set<string> = new Set();

  while (queue.length) {
    const current: string = queue.shift()!;
    if (!visited.has(current)) {
      visited.add(current);
      res.push(current);
      queue.push(
        ...graph[current].filter((neighbor) => !visited.has(neighbor))
      );
    }
  }

  return res;
};

console.log(bfs(graph, 'A')); // [A, B, C, D, E, F]
